### ISSUE TITLE
Feature Request: Routines

### ISSUE THREAD

**Issue by StaleLoafOfBread** [humbertogontijo/homeassistant-roborock#169]
Although we can replicate the routine functionality via scripts / node red / automations, I would personally prefer to "load up" the routines onto the robot and start them via home assistant. I think this would also make it a bit more beginner friendly as well. 

If anyone who knows how one might go about figuring out how to accomplish this wants to comment, I'll give it a shot. My current idea is to try to capture the traffic when using the native app. Though my precious efforts doing so to figure out local control did not go as planned so any insight is appreciated.

**Comment by Xyaren** [comment]
Sounds like a job for a HomeAssistant Script that can be called by multiple sources.
Check out the docs: https://www.home-assistant.io/integrations/script/
On your Instance: https://my.home-assistant.io/redirect/scripts/

**Comment by StaleLoafOfBread** [comment]
I appreciate the info @Xyaren! That's how I am doing it today but I would prefer to be able to use the routines as there are slight differences:
1. The pattern used when running multiple times in a row is different. Its cross hatch when a routine (well just any native way of telling it how many times)
2. Allows me to use the native app or HA to tell it to do the same thing and its kept in sync
3. When I use scripts it sometimes errors and although Im sure I could make it more robust, it would be more convient to just call a service like `roborock.routine`.

**Comment by Lash-L** [comment]
Hi @StaleLoafOfBread This would be possible - however, Routines are only available on some Roborocks. Unfortunately I don't have the functionality on mine, so this is not something I can reverse engineer. 

if you are willing to go through the work of logging your network packets (relatively easy if you have a mac and iphone) - I can attempt to add this functionality. Otherwise, I have applied the help wanted label for now.

**Comment by Regressor** [comment]
@Lash-L I have robo with routines and wish to help to implement this future. I'm a programmer and have an expierence with HA integrations.

**Comment by humbertogontijo** [comment]
Weird enough the robot only have a command to retrieve routine ids. The app stores the information about the scheduled time. Meaning that even if the integration creates one, when the app is opened and it doesn't recognize this new routine it will clear it.
We could store the routine in HA, but this means that HA is the source of true, overriding anything done in the app.
The command to create routines receives a cron, the fan speed and a command to execute (So far I only seem "start_clean"). I can't even relate this to HA entities (calendar could be the closest one, but it would require the description to be well formatted to receive the fan speed)

**Comment by Regressor** [comment]
I need an ability to load routines and execute them. Example: I created a routine to make 2 pass - first to clean and second to mop with special options for rooms. I want to be able to get list with ids and to start them via HA service.

**Comment by humbertogontijo** [comment]
@Regressor Correct if I'm wrong, but there is no such thing in the app. Just schedules. You can set it to clean predefined rooms, but not the number of cleanings and not immediately (it's scheduled).
It is achievable with HA automations though

**Comment by Regressor** [comment]
No. Routines is a sequence of operations/jobs. Every job can be set to clean or mop with separate settings (like number of cleanings, intensivity and other). Jobs are running one by one. Routine can be set to run by timer.

The main problem with HA automations is a need to setup all clean/mop parameters and rooms by hands and to run multiple jobs I need to write some code/config to detect each job ending time. With routines I can setup all parameters in application and then just run routine from HA using voice assistant ("please clean kitchen" or "do full clean").

![routines](https://github.com/humbertogontijo/homeassistant-roborock/assets/17557802/4bc5eb14-d7e7-4f77-9975-f57cd3d85c43)

**Comment by humbertogontijo** [comment]
@Regressor I managed to intercept some commands from my app, but I think those are only for creating a task. Maybe this created task is linked to a routine using mqtt

**Comment by Regressor** [comment]
> @Regressor I managed to intercept some commands from my app, but I think those are only for creating a task. Maybe this created task is linked to a routine using mqtt

How do you intercept commands ? I can try to inspect my application

**Comment by humbertogontijo** [comment]
You'll need a mac and an iPhone
Using cli from https://github.com/humbertogontijo/python-roborock with parameters `parser --device_ip <device_ip> --local_key <local_key>`
You can get the local key from the logs or from /config/.storage/core.config_entries file
Then run "rvictl -s XXXXXXXX-XXXXXXXXXXXXXXXX" with your phone connected with a USB cable to your mac, replacing XXX...XXX with your phone ID

**Comment by Lash-L** [comment]
Someone was able to get it to work for them on Windows, but I can't remember who or for what issue. If you don't have a mac/iphone, let us know and I can try to find the comment from them

**Comment by Regressor** [comment]
> Someone was able to get it to work for them on Windows, but I can't remember who or for what issue. If you don't have a mac/iphone, let us know and I can try to find the comment from them

Yes. I don't have iphone && mac.

**Comment by Lash-L** [comment]
Check here https://github.com/humbertogontijo/homeassistant-roborock/issues/298

And/or @PyRowMan can give some insight into what he did

**Comment by Regressor** [comment]
> Check here #298
Thank you. I'll try to use this.

**Comment by PyRowMan** [comment]
Hi There, sorry for the long time to reply, 

@Regressor You can block roborock dns with adblock dns and then use PCAPDroid to intercept the outgoing PCAP from roborock app :)

**Comment by StaleLoafOfBread** [comment]
@PyRowMan Did you need to use the rooted version?

**Comment by StaleLoafOfBread** [comment]
@PyRowMan @humbertogontijo  I finally rooted my phone but now when I block the DNS or unplug my modem, the app doesn't work at all. Can you confirm that you are still able to get the app working in local mode?

**Comment by StaleLoafOfBread** [comment]
I changed my DNS blocking from 
```
*.roborock.com
*.xiaomi.com
```

to just `mqtt-us-2.roborock.com`

but I suspect this is not working as expected. Here is what the parser shows
![image](https://github.com/humbertogontijo/homeassistant-roborock/assets/45444205/6bfca181-f49b-460e-b546-5101d040eef0)

**Comment by Regressor** [comment]
> Hi There, sorry for the long time to reply,
> 
> @Regressor You can block roborock dns with adblock dns and then use PCAPDroid to intercept the outgoing PCAP from roborock app :)

Sorry for long delay. I was in business trip :(
I created an emulator android device (android studio native one) and installed roborock app, pcapdroid and adblock. But. roborock app use TLS connections and don't trust external CA certificates, so pcapdroid can't decode tls.

**Comment by StaleLoafOfBread** [comment]
> Sorry for long delay. I was in business trip :( I created an emulator android device (android studio native one) and installed roborock app, pcapdroid and adblock. But. roborock app use TLS connections and don't trust external CA certificates, so pcapdroid can't decode tls.

No worries. With pcapandroid you can use a MITM cert to decode it though, can't you? I know it has the feature but this is the first I've tried using this app or doing this sort of work with android apps. 

What did you block with adbock?

**Comment by Regressor** [comment]
MITM doesn't work with roborock app. I can use MITM with Chrome - and it work and TLS session dumped decoded, but roborock just don't trust pcapdroid CA:

![image](https://github.com/humbertogontijo/homeassistant-roborock/assets/17557802/973d225a-25f3-4c4f-b563-b6507add6481)

About adblock. I didn't use it. I just can't find "adblock dns" app that can block dns. Play market on "adblock dns" shows many apps: ADGuard, Purple DNS, Wize Adblock VPN, FAB Adblocker browser, Adblock browser, Adguard vpn and there is no "adblock dns" app.

But why blocking dns can solve tls problem ?

**Comment by StaleLoafOfBread** [comment]
@Regressor  My goal here is to capture the local traffic when the app starts a routine. It was my understanding the way to get it to start talking local was to block certain dns queries. I see now from rereading the thread that it was @PyRowMan who got it to work on Android.

**Comment by Regressor** [comment]
I think pcap droid installs vpn with non-local address and this can prevent roborock app send local traffic. But I'll try this when find adblock app

**Comment by StaleLoafOfBread** [comment]
> I think pcap droid installs vpn with non-local address and this can prevent roborock app send local traffic. But I'll try this when find adblock app.

It does unless you are using the rooted version. 

> Can I disable internet for roborock app and capture app <-> robot traffic on mikrotik router ?

I am unfamiliar with that router brand but thats the general idea. However, I was having issues with that, which is why I was asking what dns blocking help worked.

**Comment by Regressor** [comment]
> It does unless you are using the rooted version.
Oh. It does matter.

> I am unfamiliar with that router brand but thats the general idea. However, I was having issues with that, which is why I was asking what dns blocking help worked.
First try failed. I was able to setup robot local traffic capture (mikrotik TZSP streaming to pc with wireshark). Also I added firewall rule to block internet for roborock app phone. And there is no any local traffic at all

**Comment by StaleLoafOfBread** [comment]
That's roughly the results I saw. I can't remember the exact details anymore. I was hoping someone who successfully captured the data could share their technique.

**Comment by Regressor** [comment]
Ok. Second try was successfull, but it looks like this technique will not help:

1. I opened roborock application and touched "Enter" button.
2. After loading map I disabled internet for phone on router
3. Then I setup a packet sniffer streaming on wifi router to computer with wireshark running
4. Then I can start/stop cleaning, changing options and local traffic is captured

But there is a problem - Routines are placed under "Enter" button and they don't work without internet. I think routines are cloud only option and only way to intercept and decode traffic is to decompile and change roborock app.

**Comment by StaleLoafOfBread** [comment]
Looks like @humbertogontijo had [some sort of success](https://github.com/humbertogontijo/homeassistant-roborock/issues/419#issuecomment-1646564627 ) with the routines though he was creating routines where my (our?) goal is to just start the routines we made in app. 

I'll have to test if I can get routines working offline later as I can't remember. I thought I did but based off what you just said, I must be misremembering. I don't have a good way to do that right now without killing internet for all devices which just isn't an option right now.

**Comment by Regressor** [comment]
> Looks like @humbertogontijo had [some sort of success](https://github.com/humbertogontijo/homeassistant-roborock/issues/419#issuecomment-1646564627) with the routines though he was creating routines where my (our?) goal is to just start the routines we made in app.

Yes. My goal is to be able to run routines. I like the way I can create them from app and want to run routines with voice assistant.

**Comment by Regressor** [comment]
It's a shame, but apparently the routines function is completely cloud-based. I ran a routine consisting of two tasks (vacuum and then wash) and immediately after that I turned off the Internet for the robot. The vacuum cleaner completed the first task, docked and did not complete the second part even after turning on the Internet.

I'll try to make an automation for this in HA without routines...

**Comment by muinmomin** [comment]
> Hi @StaleLoafOfBread This would be possible - however, Routines are only available on some Roborocks. Unfortunately I don't have the functionality on mine, so this is not something I can reverse engineer.
> 
> if you are willing to go through the work of logging your network packets (relatively easy if you have a mac and iphone) - I can attempt to add this functionality. Otherwise, I have applied the help wanted label for now.

@Lash-L -- Was able to look at the requests. Routines are "scenes" in the API. You can execute a scene with `https://api-us.roborock.com/user/scene/<scene-id>/execute`

Scenes are fetched with `https://api-us.roborock.com/user/scene/device/<device id>`

**Comment by Lash-L** [comment]
@muinmomin can you please give me an example result of getting the scenes?

**Comment by muinmomin** [comment]
@Lash-L sure:

```
{
"api": null,
"result": [
    {
      "id": <id here>,
      "name": "Full",
      "param": "{\"triggers\":[],\"action\":{\"type\":\"S\",\"items\":[{\"id\":1,\"type\":\"CMD\",\"name\":\"\",\"entityId\":\"<redacted>\",\"param\":\"{\\\"id\\\":1,\\\"method\\\":\\\"do_scenes_app_start\\\",\\\"params\\\":[{\\\"fan_power\\\":104,\\\"water_box_mode\\\":203,\\\"mop_mode\\\":300,\\\"mop_template_id\\\":300,\\\"repeat\\\":1,\\\"auto_dustCollection\\\":1,\\\"source\\\":101}]}\",\"finishDpIds\":[130]},{\"id\":2,\"type\":\"CMD\",\"name\":\"Kitchen\",\"entityId\":\"<redacted>\",\"param\":\"{\\\"id\\\":2,\\\"method\\\":\\\"do_scenes_segments\\\",\\\"params\\\":{\\\"data\\\":[{\\\"tid\\\":\\\"<redacted>\\\",\\\"segs\\\":[{\\\"sid\\\":16}],\\\"map_flag\\\":0,\\\"fan_power\\\":102,\\\"water_box_mode\\\":202,\\\"mop_mode\\\":300,\\\"mop_template_id\\\":300,\\\"repeat\\\":1,\\\"clean_order_mode\\\":1,\\\"auto_dry\\\":1,\\\"auto_dustCollection\\\":1,\\\"region_num\\\":0}],\\\"source\\\":101}}\",\"finishDpIds\":[130]}]},\"matchType\":\"NONE\"}",
      "enabled": true,
      "extra": null,
      "type": "WORKFLOW"
    }],
  "status": "ok",
  "success": true
}
```

**Comment by muinmomin** [comment]
Request for getting scenes
```
curl 'https://api-us.roborock.com/user/scene/device/<device id>' \
-H 'Host: api-us.roborock.com' \
-H 'Accept: */*' \
-H 'Authorization: ... \
-H 'X-APP-NAME: com.roborock.smart' \
-H 'X-UID: ...' \
-H 'Accept-Language: en-US;q=1' \
-H 'X-APP-VERSION-CODE: 4.42.02' \
-H 'X-IOTSDK-VERSION: 3.2.0_11_12' \
-H 'User-Agent: com.roborock.smart/4.42.02_1 (iPhone16,1; iOS 18.2; Scale/3.00)' \
-H 'Connection: keep-alive' \
--cookie '...' \
```

Request for triggering a scene
```
curl 'https://api-us.roborock.com/user/scene/<scene id>/execute' \
-X POST \
-H 'Host: api-us.roborock.com' \
-H 'X-IOTSDK-VERSION: 3.2.0_11_12' \
-H 'Accept: */*' \
-H 'Authorization: ... \
-H 'X-APP-NAME: com.roborock.smart' \
-H 'X-UID: ... \
-H 'Accept-Language: en-US;q=1' \
-H 'X-APP-VERSION-CODE: 4.42.02' \
-H 'User-Agent: com.roborock.smart/4.42.02_1 (iPhone16,1; iOS 18.2; Scale/3.00)' \
-H 'Connection: keep-alive' \
-H 'X-Locale: eyJ0aW1lem9uZSI6IkFtZXJpY2FcL0NoaWNhZ28iLCJzaW0iOjB9' \
-H 'Content-Type: application/x-www-form-urlencoded' \
--cookie '...' \
```

**Comment by regevbr** [comment]
please add a thumbs up for https://github.com/Python-roborock/python-roborock/pull/317 to get this rolling

**Comment by StaleLoafOfBread** [comment]
I copied the version from the current dev version in home assistant and added it to mine as a custom integration to test out the new feature. Works great! https://github.com/home-assistant/core/tree/7e4432e321cedeedbfe4d7a45eb40b7078e5f5e9/homeassistant/components/roborock

(I did have to change `AddConfigEntryEntitiesCallback` back to `AddEntitiesCallback` to make this work since I'm running `2025.2.5`)

Since I run the official integration now, as far as I'm concerned, this is completed as the code is in the official integration and assumingly will come out soon ([pr](https://github.com/home-assistant/core/pull/137203) wasn't tagged with `2025.3.0` so unsure if it's coming out next week or not). Leaving it up to your discretion @humbertogontijo @Lash-L if this issue should remain open till its officially released / till it makes it into this actual repo as I don't see a PR for this repo. 

Thank you so much to everyone who made this possible! Now if only Roborock made this feature support offline...

**Comment by StaleLoafOfBread** [comment]
Actually quick question for @regevbr, why did you use `scene` instead of `button`?

**Comment by regevbr** [comment]
> Actually quick question for [@regevbr](https://github.com/regevbr), why did you use `scene` instead of `button`?

One of the HA maintainers insisted it is the correct integration, I disagree, but who am I to determine :-)

**Comment by StaleLoafOfBread** [comment]
> One of the HA maintainers insisted it is the correct integration, I disagree, but who am I to determine :-)

Don't know how I missed [that](https://github.com/home-assistant/core/pull/137203#issuecomment-2629603992) originally. Thank you for the reply! FWIW I agree with you that buttons is correct but I too am not one to make such decisions.

**Comment by StaleLoafOfBread** [comment]
@regevbr 
Just hopped on to do my best to restore your button work and saw https://github.com/home-assistant/core/pull/139845. That's locked so wanted to comment on here to say thank you! I appreciate your efforts very much!
